base/iterators.jl

```diff
@@ -1128,6 +1128,7 @@ end
 reverse(p::ProductIterator) = ProductIterator(Base.map(reverse, p.iterators))
 last(p::ProductIterator) = Base.map(last, p.iterators)
 intersect(a::ProductIterator, b::ProductIterator) = ProductIterator(intersect.(a.iterators, b.iterators))
+getindex(p::ProductIterator, inds...) = map(getindex, p.iterators, inds)
```

Comment:
map returns an array right? it should probably return a tuple

Reply:
probably something like
```julia
function getindex(prod::ProductIterator, indices...)
    return _prod_getindex(prod.vectors, indices...)
end
_prod_getindex(::Tuple{}) = ()
function _prod_getindex(p_vecs::Tuple, indices...)
    return (
        first(prod.vectors)[first(indices)],
        _prod_getindex(Base.tail(p_vecs), Base.tail(indices)...)
    )
end
```

Reply:
judging from the code above

Reply:
No, that's not right. `map` has methods that produce `Tuple`s. 
```julia
julia> map(getindex, (['a', 'b'], 5:6), (2, 1))
('b', 5)
```

Reply:
huh interesting - how does it decide which one?

Reply:
Just mutliple dispatch. https://github.com/JuliaLang/julia/blob/master/base/tuple.jl#L320

If the inputs are tuples then you know the size of the outputs, so it can map Tuples to Tuples.